xarray/core/dataarray.py

```diff
@@ -64,6 +65,8 @@
 )
 
 if TYPE_CHECKING:
+    T_DSorDA = TypeVar("T_DSorDA", "DataArray", Dataset)
```

Comment:
Should this be a Union?

Reply:
(Disclaimer, I don't know the type system that you all are using, and I don't have much expeience using Python type annotations).

Reply:
This is a @crusaderky question.

Reply:
No. The whole point of this contrived syntax (which I hate, but I don't think there's any other sensible way) is to allow to define

```python
def f(x: T_DSorDA) -> T_DSorDA
```
which is the same as writing
```python
@overload
def f(x: DataArray) -> DataArray:
    ...

@overload
def f(x: Dataset) -> Dataset:
    ...

def f(x):
    # implementation
```
except that it's a lot more compact, particularly when other arguments start being involved (like in this case), and doesn't explode cartesian-ly (is it a word? :P) when two or more separate overloading combinations enter need to be represented.

In constrast,
```python
def f(x: Union[DataArray, Dataset]) -> Union[DataArray, Dataset]
```
means that the output of the function is unrelated with the input, which causes problems down the line when you use the function; e.g.
```python
x = DataArray()
y = f(x)
y[0] # mypy error: Dataset is not subscriptable
```

If instead you were to use a simpler, unbound TypeVar, then mypy would complain in the implementation of the function:
```python
T = TypeVar("T")

def f(x: T) -> T:
    return x.copy()  # mypy error: object has no attribute "copy"

x = DataArray()
y = f(x)
y[0] # this one instead is ok; mypy knows that y is a DataArray
```

Reply:
Forgot to mention that the last syntax completely disables validation on the parameters;
``y = f(123)`` won't raise any errors.

Reply:
That makes sense.  Thanks for the explanation @crusaderky !